modules/lldp/sink.c

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include "gr_lldp.h"
+#include "lldp_priv.h"
+
+#include <gr_eth_input.h>
+#include <gr_eth_output.h>
+#include <gr_graph.h>
+#include <gr_ip4_control.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+#include <gr_port.h>
+
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+
+#include <time.h>
+
+enum {
+	LLDP_REPLY = 0,
+	DROP_IN,
+	EDGE_COUNT,
+};
+extern struct gr_lldp_node_data lldp_ctx;
+
+struct gr_lldp_neigh neighbors[RTE_MAX_ETHPORTS];
+
+static uint16_t
+lldp_sink_process(struct rte_graph *, struct rte_node *, void **objs, uint16_t nb_objs) {
+	struct iface_info_port *port;
+	const struct iface *iface;
+	struct rte_mbuf *mbuf;
+	clock_t now = clock();
+
+	if (lldp_ctx.rx) {
+		for (uint16_t i = 0; i < nb_objs; i++) {
+			mbuf = objs[i];
+			iface = eth_input_mbuf_data(mbuf)->iface;
+			port = (struct iface_info_port *)(iface->info);
+
+			neighbors[port->port_id].last_seen = now;
+			neighbors[port->port_id].iface_id = iface->id;
+			neighbors[port->port_id].n_tlv_data = rte_pktmbuf_data_len(mbuf);
+			memcpy(neighbors[port->port_id].tlv_data,
+			       rte_pktmbuf_mtod(mbuf, void *),
+			       rte_pktmbuf_data_len(mbuf));
+		}
+	}
+
+	rte_pktmbuf_free_bulk((struct rte_mbuf **)objs, nb_objs);
+	return nb_objs;
+}
+
+static void lldp_sink_register(void) {
+	rte_edge_t edge = gr_node_attach_parent("eth_input", "lldp_sink");
+	if (edge == RTE_EDGE_ID_INVALID)
+		ABORT("gr_node_attach_parent(eth_input, lldp_input) failed");
+	gr_eth_input_add_type(rte_cpu_to_be_16(RTE_ETHER_TYPE_LLDP), edge);
+}
+
+static void lldp_sink_unregister(void) { }
+
+static struct rte_node_register node = {
+	.name = "lldp_sink",
```

Comment:
I'm not familiar with LLDP. Is there a reason why this node is called lldp_sink instead of lldp_input?

Reply:
No related to lldp, but I wanted to express that the packets won't be forwarded.
Won't `*_input`  imply that there will be subsequent nodes ?

Reply:
OK, I would still prefer if we kept the input/output terminology to be consistent with other nodes. Since the node names used when displaying software stats, it will be less confusing.